### PR TITLE
fix: Remove validation of `super` in use paths

### LINKED ISSUE #10770: False Syntax Error: The `super` keyword may only be preceded by other `super`s
There's a bug in the syntax parser.

### How to reproduce?

Create an arbitrary library crate with `cargo new my_project --lib` then edit `src/lib.rs` as follows:
```rust
pub fn main() {
    println!("Hello, world!");
}

pub mod submodule {
    use self::super::main;

    pub fn foo() {
        main();
    }
}
```

### What's expected?

No diagnostics shown.  Indeed, `cargo check` (and `cargo clippy`) shows nothing.

### What happens?

The following error is detected on "super" in the 6th line:
```
[rust-analyzer syntax-error] [E] Syntax Error: The `super` keyword may only be preceded by other `super`s
```

### How to workaround?

Replace `self::super` with `super`.  Perhaps they're identical (I'm not quite sure).

### PR DESCRIPTION
**Pull request by lnicola**
Fixes #10770

### CODE REVIEW

File: crates/syntax/src/validation.rs [hunk 1]
```diff
@@ -264,26 +264,6 @@ fn validate_path_keywords(segment: ast::PathSegment, errors: &mut Vec<SyntaxErro
...
-        if segment.coloncolon_token().is_some() || !all_supers(&path) {
-            errors.push(SyntaxError::new(
-                "The `super` keyword may only be preceded by other `super`s",
-                token.text_range(),
-            ));
-            return;
-        }
-
-        let mut curr_path = path;
-        while let Some(prefix) = use_prefix(curr_path) {
```
Comment (lnicola) [inline crates/syntax/src/validation.rs:277]: This looks weird (it does O(N) passes over the path).

**Comment by lnicola** [comment]
This still allows e.g. `use std::io::copy::super`, but I don't think we can spot that `copy` is a function and not a module at this level.

**Comment by Veykril** [comment]
> This still allows e.g. `use std::io::copy::super`, but I don't think we can spot that `copy` is a function and not a module at this level.

A diagnostic for that probably has to happen in name resolution?
bors r+

**Comment by lnicola** [comment]
PS: the numbered tests are a bit annoying

**Pull request merged**